scripts/python/golioth/golioth/golioth.py

```diff
@@ -0,0 +1,363 @@
+from __future__ import annotations
+
+from contextlib import asynccontextmanager
+from datetime import datetime, timedelta, timezone
+from enum import Enum, IntEnum
+import json
+from pathlib import Path
+import re
+from typing import Any, Iterable, Optional, Union
+
+import httpx
+import yaml
+from trio_websocket import open_websocket_url, WebSocketConnection
+
+
+class ProjectNotFound(RuntimeError):
+    pass
+
+
+class DeviceNotFound(RuntimeError):
+    pass
+
+
+class RPCStatusCode(IntEnum):
+    OK = 0
+    CANCELED = 1
+    UNKNOWN = 2
+    INVALID_ARGUMENT = 3
+    DEADLINE_EXCEEDED = 4
+    NOT_FOUND = 5
+    ALREADYEXISTS = 6
+    PERMISSION_DENIED = 7
+    RESOURCE_EXHAUSTED = 8
+    FAILED_PRECONDITION = 9
+    ABORTED = 10
+    OUT_OF_RANGE = 11
+    UNIMPLEMENTED = 12
+    INTERNAL = 13
+    UNAVAILABLE = 14
+    DATA_LOSS = 15
+    UNAUTHENTICATED = 16
+
+
+class RPCError(RuntimeError):
+    pass
+
+
+class RPCResultError(RPCError):
+    def __init__(self, status_code: int):
+        self.status_code: RPCStatusCode = RPCStatusCode(status_code)
+        super().__init__(f'RPC failed with status code {repr(self.status_code)}')
+
+
+class RPCTimeout(RPCError):
+    def __init__(self):
+        super().__init__(f'RPC timeout')
+
+
+class ApiNodeMixin:
+    @property
+    def http_client(self):
+        return httpx.AsyncClient(base_url=self.base_url)
+
+    async def get(self, *args, **kwargs):
+        async with self.http_client as c:
+            return await c.get(*args, **kwargs)
```

Comment:
It's okay to do in a future PR, but I would like to see support for using `https` with Golioth prod (or dev). It shouldn't be much more additional work, other than adding the `s` in the URL and setting the `x-api-key` header in request.

Reply:
Sure. Didn't want to spend time on something "just in case", as I use 99% self-hosted server for testing. And this tool was developed for testing primarily, with other use-cases just kept in mind.